thrust/thrust/iterator/detail/tabulate_output_iterator.inl

```diff
@@ -0,0 +1,72 @@
+// SPDX-FileCopyrightText: Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+
+#include <thrust/detail/config.h>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/iterator_adaptor.h>
+#include <thrust/iterator/tabulate_output_iterator.h>
+
+THRUST_NAMESPACE_BEGIN
+
+template <typename BinaryFunction, typename System, typename DifferenceT>
+class tabulate_output_iterator;
+
+namespace detail
+{
+
+// Proxy reference that invokes a BinaryFunction with the index of the dereferenced iterator and the assigned value
+template <typename BinaryFunction, typename DifferenceT>
+class tabulate_output_iterator_proxy
+{
+public:
+  _CCCL_HOST_DEVICE tabulate_output_iterator_proxy(BinaryFunction fun, DifferenceT index)
+      : fun(fun)
+      , index(index)
+  {}
+
+  _CCCL_EXEC_CHECK_DISABLE
+  template <typename T>
+  _CCCL_HOST_DEVICE tabulate_output_iterator_proxy operator=(const T& x)
+  {
+    fun(index, x);
+    return *this;
+  }
+
+private:
+  BinaryFunction fun;
+  DifferenceT index;
+};
+
+// Compute the iterator_adaptor instantiation to be used for tabulate_output_iterator
+template <typename BinaryFunction, typename System, typename DifferenceT>
+struct tabulate_output_iterator_base
+{
+  using type =
+    thrust::iterator_adaptor<tabulate_output_iterator<BinaryFunction, System, DifferenceT>,
+                             counting_iterator<DifferenceT>,
+                             thrust::use_default,
+                             System,
+                             thrust::use_default,
+                             tabulate_output_iterator_proxy<BinaryFunction, DifferenceT>>;
+};
```

Comment:
Suggestion: This could avoid a template instantation by just becoming an alias:
```suggestion
template <typename BinaryFunction, typename System, typename DifferenceT>
using tabulate_output_iterator_base =
    thrust::iterator_adaptor<tabulate_output_iterator<BinaryFunction, System, DifferenceT>,
                             counting_iterator<DifferenceT>,
                             thrust::use_default,
                             System,
                             thrust::use_default,
                             tabulate_output_iterator_proxy<BinaryFunction, DifferenceT>>;
```
The iterator then inherits:
```c++
class tabulate_output_iterator : public detail::tabulate_output_iterator_base<BinaryFunction, System, DifferenceT>
```

Reply:
Sounds good, I switched to using alias template instead.